lib/core/networking/protocol/project_update.dart

```diff
@@ -0,0 +1,101 @@
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:flyde/core/fs/configs/compiler_config.dart';
+import 'package:flyde/core/fs/wrapper/source_file.dart';
+import 'package:json_annotation/json_annotation.dart';
+
+part 'project_update.g.dart';
+
+/// Updates a project with new files and configuration.
+@JsonSerializable()
+class ProjectUpdateRequest {
+  /// The file list of the project.
+  ///
+  /// Each file id is mapped to the hash of it's content.
+  final Map<String, String> files;
+
+  /// The configuration which should be used for compilation.
+  final CompilerConfig config;
+
+  ProjectUpdateRequest({required this.files, required this.config});
+
+  factory ProjectUpdateRequest.fromJson(Map<String, dynamic> json) =>
+      _$ProjectUpdateRequestFromJson(json);
+
+  Map<String, dynamic> toJson() => _$ProjectUpdateRequestToJson(this);
+}
+
+/// The response to a [ProjectUpdateRequest].
+@JsonSerializable()
+class ProjectUpdateResponse {
+  /// List of files which need to be updated.
+  final List<String> files;
+
+  ProjectUpdateResponse({required this.files});
+
+  factory ProjectUpdateResponse.fromJson(Map<String, dynamic> json) =>
+      _$ProjectUpdateResponseFromJson(json);
+
+  Map<String, dynamic> toJson() => _$ProjectUpdateResponseToJson(this);
+}
+
+/// A file update message used to transfer files from the user to the server.
+@JsonSerializable()
+class FileUpdate {
+  /// Mirror of `SourceFile->name`
```

Comment:
`FileUpdate` mirrors the critical fields of `SourceFile` which are required to recreated the later one.

We cannot simply make `SourceFile` JSON serializable because `SourceFile` can keep it's data on disk or in memory.

Reply:
I got that, but the word ‘Mirror’ is a little confusing, why not just remove that, the rest would be self explanatory?